src/st/cat/unk_3872C.c

```diff
@@ -3,7 +3,83 @@
 
 INCLUDE_ASM("st/cat/nonmatchings/unk_3872C", func_us_801B872C);
 
-INCLUDE_ASM("st/cat/nonmatchings/unk_3872C", func_us_801B87E8);
+void func_us_801B87E8(Entity* self) {
+    Primitive* prim;
+    s32 primIndex;
+    s16 base_x;
+    s32 cntr;
+
+    for (base_x = -g_Tilemap.scrollX.i.hi; base_x < (-0x3E); base_x += 0x3E)
+        ;
+
+    if (self->step == 0) {
+        self->step += 1;
+        do {
+            primIndex =
+                (self->primIndex = g_api_AllocPrimitives(PRIM_GT4, 0x10));
+        } while (0);
+        if (primIndex == (-1)) {
+            DestroyEntity(self);
+            return;
+        }
+        self->flags = FLAG_KEEP_ALIVE_OFFCAMERA | FLAG_HAS_PRIMS;
+        prim = &g_PrimBuf[self->primIndex];
+        cntr = 0;
+        while (prim != 0) {
+            prim->x0 = prim->x2 = base_x + ((cntr & 0x7) * 0x3E);
+            prim->x1 = prim->x3 = prim->x0 + 0x3E;
+            prim->y1 = prim->y0 = ((cntr >> 3) * 0x5F) + 0x1C;
+            prim->tpage = 0xF;
+            prim->clut = 0x8F;
+            prim->priority = 0x20;
+            prim->u0 = 1;
+            prim->v0 = 0x81;
+            prim->u1 = 0x3F;
+            prim->v1 = 0x81;
+            prim->u2 = 1;
+            prim->v2 = 0xDE;
+            prim->u3 = 0x3F;
+            prim->v3 = 0xDE;
+            prim->drawMode = DRAW_UNK02;
+            prim->y3 = prim->y2 = ((u16)prim->y0) - (-0x5F);
+            if (g_CastleFlags[0x42] == 0) {
+                prim->drawMode = DRAW_HIDE | DRAW_COLORS | DRAW_UNK02;
+            }
+            prim = prim->next;
+            cntr++;
+        }
+
+    } else {
+        cntr = 0;
+        if (self->ext.ILLEGAL.u8[4] != g_CastleFlags[0x42]) {
+            self->ext.ILLEGAL.u8[5] = 01;
+        }
+        prim = &g_PrimBuf[self->primIndex];
+        while (prim != 0) {
+            if (g_CastleFlags[0x42] != 0) {
+                prim->drawMode &= ~DRAW_HIDE;
+            }
+            if (self->ext.ILLEGAL.u8[5] != 0) {
+                if (prim->r0 > 0x7f) {
+                    self->ext.ILLEGAL.u8[5] = 0;
+                } else {
+                    prim->r0 = prim->r0 + 2;
+                    prim->g0 = (prim->b0 = prim->r0);
+                    *(u32*)&prim->r1 = *(u32*)&prim->r0;
+                    *(u32*)&prim->r2 = *(u32*)&prim->r0;
+                    *(u32*)&prim->r3 = *(u32*)&prim->r0;
```

Comment:
These three lines are duplicated everywhere in the codebase
```suggestion
                    LOW(prim->r1) = LOW(prim->r0);
                    LOW(prim->r2) = LOW(prim->r0);
                    LOW(prim->r3) = LOW(prim->r0);
```